s1m0nWasTaken/CS4337_BigDataProject
# Role-based authorisation for all microservices

## Summary
I added role-based authorisation for every microservice using shared request filters and JWT utilities.

This works so that if you log in and use your JWT and refresh token you will be allowed to access endpoints based on your role. Endpoint behaviour may also vary depending on your role and user id (e.g. only admins can ban, only the same user or an admin can update their profile, etc.).

JwtRequestFilter validates JWTs with the help of JwtUtils and adds user id and role to the SecurityContextHolder. It also will try refresh a JWT if it's expired and if the refresh token is valid, the new JWT can be found in the response header you get.

ApiKeyAuthFilter allows any requests that are using the API key and gives the context the admin role.

I have added a special restTemplate for each service with the "authRestTemplate" qualifier as this will be used for refreshing the JWT when needed. Rest templates can be found in Config or AppConfig.

Any rest templates used for cross-service communication now has an interceptor that adds the API key. This interceptor is ApiKeyInterceptor.

What I have yet to do/am struggling with:
- Abstract some common functions to help with authorisation to a shared class called SecurityContextHolderUtils.
- I can't figure out why @ Value doesn't work for the API key in all ApiKeyInterceptors. -- **FIXED**
- For some reason ApiKeyInterceptor doesn't seem to work when I try to reuse it in SharedObjs.
- I realise I have leaked the API key so I will change it once I figure out how to use @ Value in ApiKeyInterceptors. -- **FIXED**
- Out of scope but obviously the unit tests need to be fixed. -- **FIXED**

If anyone has any ideas on the above please share.

## How to verify
**Make sure the necessary services are up and running when testing (Auth is always needed)!**

**Because I broke some tests if a service doesn't build just comment out the tests for now**

Unfortunately you have to test every endpoint to thoroughly confirm if this is working as intended. As a compromise I would appreciate if each of us could test the services we worked on ourselves and that way we can cover everything.

You will need a valid JWT to hit the endpoints (or a refresh token with an expired token and then you can get the refreshed JWT in the response headers). You can add JWTs like this in Postman:
![image](https://github.com/user-attachments/assets/2ea8ec1a-a2a9-4e26-86c3-ed2bc640cdfb)

Here is an example of the refreshed JWT in the response headers:
<img width="736" alt="image" src="https://github.com/user-attachments/assets/c01568eb-9bdb-40a4-830b-fc067d9c95da">

You can also test requests using our API key which is used for microservices to bypass authentication when communicating with each other but since this is limited to a few requests, it's better to focus on testing the roles (requests send using the API key are essentially the same as an admin making requests).
![image](https://github.com/user-attachments/assets/db0ae684-b4b8-46da-909f-93707fe1a7c6)

If you want to test endpoints using a different role you can update your role using this endpoint easily with the API key:
![image](https://github.com/user-attachments/assets/10319eea-f0e2-4d06-a723-3addce3583f6)

Be sure to look out for how I configured the authorisation of endpoints. We don't have it set in stone so I did what made the most sense to me, but if you think a certain role should have access to an endpoint please leave a comment.

**Don't use both the JWT and API key in the same request!**

## Tests
I actually broke a bunch of tests... Fixed them :D (Alex helped too)

## Related
N/A
